IOPool/Input/src/RepeatingCachedRootSource.cc

```diff
@@ -0,0 +1,368 @@
+// -*- C++ -*-
+//
+// Package:     Subsystem/Package
+// Class  :     RepeatingCachedRootSource
+//
+// Implementation:
+//     [Notes on implementation]
+//
+// Original Author:  Chris Jones
+//         Created:  Mon, 15 Mar 2021 19:02:31 GMT
+//
+
+// system include files
+#include <memory>
+
+// user include files
+#include "DataFormats/Provenance/interface/ProductRegistry.h"
+#include "DataFormats/Provenance/interface/ProcessHistoryRegistry.h"
+#include "DataFormats/Provenance/interface/BranchIDListHelper.h"
+#include "DataFormats/Provenance/interface/ProcessConfiguration.h"
+#include "DataFormats/Provenance/interface/ThinnedAssociationsHelper.h"
+#include "DataFormats/Common/interface/WrapperBase.h"
+
+#include "IOPool/Common/interface/RootServiceChecker.h"
+
+#include "FWCore/Framework/interface/InputSource.h"
+#include "FWCore/Framework/interface/ProductSelectorRules.h"
+#include "FWCore/Framework/interface/DelayedReader.h"
+#include "FWCore/Framework/interface/InputSourceDescription.h"
+#include "FWCore/Framework/interface/EventPrincipal.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/Catalog/interface/InputFileCatalog.h"
+#include "FWCore/Utilities/interface/propagate_const.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/Sources/interface/EventSkipperByID.h"
+
+#include "FWCore/Framework/interface/InputSourceMacros.h"
+
+#include "RunHelper.h"
+#include "RootFile.h"
+#include "InputFile.h"
+#include "DuplicateChecker.h"
+
+namespace edm {
+  class RunHelperBase;
+  class RCDelayedReader;
+
+  class RepeatingCachedRootSource : public InputSource {
+  public:
+    RepeatingCachedRootSource(ParameterSet const& pset, InputSourceDescription const& desc);
+
+    static void fillDescriptions(ConfigurationDescriptions& descriptions);
+
+    std::shared_ptr<WrapperBase> getProduct(unsigned int iStreamIndex,
+                                            BranchID const& k,
+                                            EDProductGetter const* ep) const;
+
+    class RCDelayedReader : public edm::DelayedReader {
+    public:
+      std::shared_ptr<edm::WrapperBase> getProduct_(edm::BranchID const& k, edm::EDProductGetter const* ep) final {
+        return m_source->getProduct(m_streamIndex, k, ep);
+      }
+      void mergeReaders_(edm::DelayedReader*) final { assert(false); }
```

Comment:
Will this function be ever called?

Reply:
This DelayedReader is only used on EventPrincipals and events do not do mergers. Therefore it should never be called.